changelog.d/9247.feature

```diff
@@ -0,0 +1 @@
+Add experimental support for backfilling history into rooms ([MSC2716](https://github.com/matrix-org/matrix-doc/pull/2716)).
```

Comment:
I'm confused by these mypy type linting errors: https://buildkite.com/matrix-dot-org/synapse/builds/15241#a41af862-06c9-47e3-91ea-bcaeeac72a18

They look unrelated to my changes here and I've merged the latest `develop` here 🤔

Reply:
Looks like an issue with the return type we defined on `parse_string` (from [comment](https://github.com/matrix-org/synapse/pull/9247#discussion_r573719702). Will look into this more

Reply:
I've removed the return type hint for `parse_string` to see how the tests roll out. Also not sure if there is an easy way to solve all of those cases. Seems like it would require a bit of refactoring which we should handle in another PR.

Reply:
Handling it in a separate PR sounds good to me. We should probably add type hints to all those functions in one go TBH.